azure-pipelines/tests_pipeline.yml

```diff
@@ -0,0 +1,39 @@
+trigger:
+  batch: true
+  branches:
+    include:
+    - trunk
+
+pr:
+  branches:
+    include:
+    - trunk
+
+variables:
+  INFRASTRUCTURE: "AZUREPIPELINES"
+  TRIAL_REPORTER: "verbose"
+
+jobs:
+
+- template: 'macos_test_jobs.yml'
+  parameters:
+    pythonVersions:
+      py27: "2.7"
+      py36: "3.6"
+      py37: "3.7"
+
+- template: 'windows_test_jobs.yml'
+  parameters:
+    pythonVersions:
+      py27: "2.7"
+      py36: "3.6"
+      py37: "3.7"
+    reactor: select
+
+- template: 'windows_test_jobs.yml'
+  parameters:
+    pythonVersions:
+      py27: "2.7"
+      py36: "3.6"
+      py37: "3.7"
+    reactor: iocp
```

Comment:
I don't like the `pythonVersions` duplication here at all... Is there a way to set a "constant" thing? "Parameters" here would required turning on "pipeline variables", and "Variables" appear to be exclusively strings.

Reply:
(Hot take: we should probably be generating this file programmatically in Python rather than doing programming YAML…)

Reply:
I agree entirely, I'm not happy about the "tiered" implicit matrix thing... this is overall too complicated, and the matrix is varied at multiple levels of yaml.

`azure-pipelines/run_test_steps.yml` is the core test run. This is mostly-well-factored, but doesn't describe which Azure VM it should be ran on: it's a set of "Steps", which are executed as part of a "Job", and you can only declare `pool.vmImage` at the pipeline, stage, or job level. MS would call `run_test_steps.yml` a "[Step Template](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=azure-devops&tabs=schema%2Cparameter-schema#step-templates)", confusingly (wouldn't Steps template make more sense?).

`azure-pipelines/windows_test_jobs.yml` and `azure-pipelines/macos_test_jobs.yml` handle the VM image declaration, but otherwise duplicate the "make a job for each specified python version" concern.

`azure-pipelines/tests_pipeline.yml` duplicates the python versions, and also varies the windows reactor.

## Intended Matrix

- python version: 2.7, 3.6, 3.7
- platform: mac, windows
- windows reactor: iocp, select (only on windows)

I think the best way would be to:
1. Generate the whole matrix, a la the script referenced by https://github.com/altendky/romp/blob/602fe43fda8c491f7799f338eaa4ce28ed96cd74/azure-pipelines.yml#L43 
2. Use that to produce jobs all in the root `azure-pipelines/tests_pipeline.yml`, each using the singular `azure-pipelines/run_test_steps.yml` template.
3. Eliminate `azure-pipelines/{macos,windows}_test_jobs.yml` entirely using matrix-produced parameters in the root pipeline definition.

@altendky the technique you use in romp looks workable. Care to make another ticket for adding that? :)